launcher/archive/ArchiveWriter.cpp

```diff
@@ -0,0 +1,211 @@
+            success = false;
+        }
+        if (archive_write_free(m_archive) != ARCHIVE_OK) {
+            qCritical() << "Failed to free archive" << m_filename << "-" << archive_error_string(m_archive);
+            success = false;
+        }
+        m_archive = nullptr;
+    }
+    return success;
+}
+
+bool ArchiveWriter::addFile(const QString& fileName, const QString& fileDest)
+{
+    QFileInfo fileInfo(fileName);
+    if (!fileInfo.exists()) {
+        qCritical() << "File does not exists:" << fileInfo.filePath();
+        return false;
+    }
+
+    std::unique_ptr<archive_entry, void (*)(archive_entry*)> entry_ptr(archive_entry_new(), archive_entry_free);
+    auto entry = entry_ptr.get();
+    if (!entry) {
+        qCritical() << "Failed to create archive entry";
+        return false;
+    }
+
+    auto fileDestUtf8 = fileDest.toUtf8();
+    archive_entry_set_pathname(entry, fileDestUtf8.constData());
+
+    QByteArray utf8 = fileInfo.absoluteFilePath().toUtf8();
+    const char* cpath = utf8.constData();
+    struct stat st;
+    if (stat(cpath, &st) == 0) {
+        archive_entry_copy_stat(entry, &st);
+    }
+    archive_entry_set_perm(entry, fileInfo.permissions());
+
+    if (fileInfo.isSymLink()) {
+        auto target = fileInfo.symLinkTarget().toUtf8();
+        archive_entry_set_filetype(entry, AE_IFLNK);
+        archive_entry_set_symlink(entry, target.constData());
+        archive_entry_set_size(entry, 0);
+
+        if (archive_write_header(m_archive, entry) != ARCHIVE_OK) {
+            qCritical() << "Failed to write symlink header for:" << fileDest << "-" << archive_error_string(m_archive);
+            return false;
+        }
+        return true;
+    }
+
+    if (!fileInfo.isFile()) {
+        qCritical() << "Unsupported file type:" << fileInfo.filePath();
+        return false;
+    }
+
+    QFile file(fileInfo.absoluteFilePath());
+    if (!file.open(QIODevice::ReadOnly)) {
+        qCritical() << "Failed to open file: " << fileInfo.filePath();
+        return false;
+    }
+
+    archive_entry_set_filetype(entry, AE_IFREG);
+    archive_entry_set_size(entry, file.size());
+
+    if (archive_write_header(m_archive, entry) != ARCHIVE_OK) {
+        qCritical() << "Failed to write header for: " << fileDest << "-" << archive_error_string(m_archive);
+        return false;
+    }
+
+    constexpr qint64 chunkSize = 8192;
+    QByteArray buffer;
+    buffer.resize(chunkSize);
+
+    while (!file.atEnd()) {
+        auto bytesRead = file.read(buffer.data(), chunkSize);
+        if (bytesRead < 0) {
+            qCritical() << "Read error in file: " << fileInfo.filePath();
+            return false;
+        }
+
+        if (archive_write_data(m_archive, buffer.constData(), bytesRead) < 0) {
+            qCritical() << "Write error in archive for: " << fileDest;
+            return false;
+        }
+    }
```

Comment:
And write the header for all files, directories, and symlinks above, while only writing the data for regular files here

```suggestion
    if (fileInfo.isFile() && !fileInfo.isSymLink()) {
        QFile file(fileInfo.absoluteFilePath());
        if (!file.open(QIODevice::ReadOnly)) {
            qCritical() << "Failed to open file:" << fileInfo.filePath();
            return false;
        }

        constexpr qint64 chunkSize = 8192;
        QByteArray buffer;
        buffer.resize(chunkSize);
    
        while (!file.atEnd()) {
            auto bytesRead = file.read(buffer.data(), chunkSize);
            if (bytesRead < 0) {
                qCritical() << "Read error in file: " << fileInfo.filePath();
                return false;
            }
    
            if (archive_write_data(m_archive, buffer.constData(), bytesRead) < 0) {
                qCritical() << "Write error in archive for: " << fileDest;
                return false;
            }
        }
    }
```

Reply:
here I do not need the check for a symlink as I exit early above if this happens